### PR TITLE
TW-435: Save Gallery Action

### PR DESCRIPTION
**Pull request by imGok**
### Zoom/Unzoom on double tap
zooom

### Save image 
- Android
downloademo

- IOS

https://github.com/linagora/twake-on-matrix/assets/99852347/2d9763a7-3a76-4409-aa1b-81db99eaedfc

- WEB
zoeozomwebaa

downloadwebaaa

### CODE REVIEW

File: lib/config/app_config.dart [hunk 1]
```diff
@@ -69,6 +69,7 @@ abstract class AppConfig {
   static const double borderRadius = 20.0;
   static const double columnWidth = 360.0;
   static const int fetchContactsLimit = 20;
+  static const String androidDownloadPath = '/storage/emulated/0/Download';
```
Comment (nqhhdev) [inline lib/config/app_config.dart:72]: I think it's only support for emulated Android. 

We need to implement for iOS, Android, Web

File: lib/pages/image_viewer/image_viewer.dart [hunk 2]
```diff
@@ -60,6 +75,28 @@ class ImageViewerController extends State<ImageViewer> {
     }
   }
 
+  void onDoubleTap() {
```
Comment (sherlockvn) [inline lib/pages/image_viewer/image_viewer.dart:78]: How about this widget: https://api.flutter.dev/flutter/widgets/InteractiveViewer-class.html

Comment (imGok) [reply lib/pages/image_viewer/image_viewer.dart:78]: WDYM ? We already use interactiveViewer (image_viewer_view.dart)

**Comment by drminh2807** [comment]
> Download downloadfileandroid

The loading bubble display too quick
Please add completed UI

**Comment by imGok** [comment]
> > Download downloadfileandroid
> >       
> >         [
> >           
> >         
> >             downloadfileandroid
> >           ](https://user-images.githubusercontent.com/48354990/265516160-e33f0831-c3d9-42e8-861c-1a8be0c8b6a4.gif)
> >         
> >         
> >           
> >             
> >               
> >             
> >             
> >               
> >               
> >             
> >           
> >           [
> >             
> >               
> >             
> >           ](https://user-images.githubusercontent.com/48354990/265516160-e33f0831-c3d9-42e8-861c-1a8be0c8b6a4.gif)
> 
> The loading bubble display too quick Please add completed UI

WDYM? You want a confirm dialog ?

File: lib/config/app_config.dart (hunk 1 above)
Comment (imGok) [reply lib/config/app_config.dart:72]: According to sources I found in internet (https://www.androidphonesoft.com/blog/how-to-access-storage-emulated-0-android/#What_is_emulated_storage_in_Android) it says storage emulated is available on every android not just virtual devices. I did Android and Web implementation we still need to do iOS but cannot with my setup

Comment (hoangdat) [reply lib/config/app_config.dart:72]: hmm, please use the dynamic way to get the path of download folder. Maybe you also have no permision to access this path in new Android API

**Comment by drminh2807** [comment]
> WDYM? You want a toast dialog ?




Yup, same with Telegram, but simpler

**Comment by imGok** [comment]
@drminh2807 ok, I'll put a toast. 

This PR is working for Android and Web for now

File: lib/pages/image_viewer/image_viewer_view.dart [hunk 3]
```diff
@@ -24,13 +24,16 @@ class ImageViewerView extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
     return Scaffold(
+      backgroundColor: Colors.black,
```
Comment (nqhhdev) [inline lib/pages/image_viewer/image_viewer_view.dart:27]: Pls use material color

File: lib/config/app_config.dart (hunk 1 above)
Comment (nqhhdev) [reply lib/config/app_config.dart:72]: @imGok Any update?

File: lib/pages/image_viewer/image_viewer_view.dart (hunk 3 above)
Comment (nqhhdev) [reply lib/pages/image_viewer/image_viewer_view.dart:27]: @imGok check it

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart [hunk 4]
```diff
@@ -1,49 +1,28 @@
...
-    Logs().d("MatrixFileExtension()::save()::result.error: ${result.error}");
-  }
-
-  Future<String?> getDownloadPathAndroid() async {
-    final directory = await getDownloadDirectoryAndroid();
-    if (directory == null) return null;
-    return '${directory.path}/$name';
-  }
+    if (PlatformInfos.isWeb) {
+      return saveWeb();
```
Comment (nqhhdev) [inline lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:20]: IMO: What is this, what will it do? Too difficult so understand ... Pls define explicitly, specifically the name of func

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart [hunk 5]
```diff
@@ -61,6 +40,28 @@ extension MatrixFileExtension on MatrixFile {
...
+          [bytes],
+          mimeType,
+        ),
+      ),
+    )
+      ..download = name
+      ..click();
+  }
+
+  void saveAndroid() async {
```
Comment (nqhhdev) [inline lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:56]: same

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart [hunk 6]
```diff
@@ -1,49 +1,28 @@
...
 import 'package:image_picker/image_picker.dart';
 import 'package:matrix/matrix.dart';
-import 'package:path_provider/path_provider.dart';
 import 'package:share_plus/share_plus.dart';
 import 'package:fluffychat/utils/platform_infos.dart';
 import 'package:fluffychat/utils/size_string.dart';
+import 'package:universal_html/html.dart' as html;
 
 extension MatrixFileExtension on MatrixFile {
   void save(BuildContext context) async {
```
[the remark this answers is not in the document]
Comment (imGok) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart]: renamed to downloadFile

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart [hunk 7]
```diff
@@ -61,6 +38,42 @@ extension MatrixFileExtension on MatrixFile {
...
+      "MatrixFileExtension()::downloadImageInMobile()::download on Mobile",
+    );
+
+    final result = await ImageGallerySaver.saveImage(
+      bytes ?? Uint8List(0),
+      name: name,
+    );
+
+    Fluttertoast.showToast(
+      msg: result?['isSuccess']
```
Comment (nqhhdev) [inline lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:71]: If `result is null`, what is happen?

**Comment by hoangdat** [comment]
1. [Mobile] What happens if user not grant us the permission to access storage? Android and iOS?
2. [Web] Why user can not choose folder to download?

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart (hunk 7 above)
Comment (imGok) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:71]: if result is null it will show the error toast

**Comment by imGok** [comment]
> * [Mobile] What happens if user not grant us the permission to access storage? Android and iOS?

on Android, you can still download the image in Pictures folder with no permission

**Comment by imGok** [comment]
I debugged to see why I couldn't select a directory to download. But it seems like it's browser related, you can choose if you want or not select the folder cf : https://stackoverflow.com/questions/33612566/how-to-specify-download-location-in-html-using-javascript 

browser in french sorry
demodat

**Comment by imGok** [comment]
> 
demodlnoperm

**Review by nqhhdev — APPROVED** [review]

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart (hunk 7 above)
Comment (hoangdat) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:71]: still not clear for me, what will handle the case of `result is null`?

**Review by sherlockvn — APPROVED** [review]

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart (hunk 7 above)
Comment (imGok) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:71]: the null safety will handle this case 

image

Comment (drminh2807) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:71]: it's good for me

File: lib/pages/image_viewer/image_viewer.dart [hunk 8]
```diff
@@ -25,8 +25,23 @@ class ImageViewer extends StatefulWidget {
 }
 
 class ImageViewerController extends State<ImageViewer> {
+  late TransformationController transformationController;
```
Comment (drminh2807) [inline lib/pages/image_viewer/image_viewer.dart:28]: why need `late` ?

File: lib/pages/image_viewer/image_viewer.dart [hunk 9]
```diff
@@ -60,6 +75,28 @@ class ImageViewerController extends State<ImageViewer> {
     }
   }
 
+  void onDoubleTap() {
+    final position = tapDownDetails!.localPosition;
+
+    const double scale = 3;
```
Comment (drminh2807) [inline lib/pages/image_viewer/image_viewer.dart:81]: you should define it at the top of controller

File: lib/pages/image_viewer/image_viewer.dart [hunk 10]
```diff
@@ -60,6 +75,28 @@ class ImageViewerController extends State<ImageViewer> {
     }
   }
 
+  void onDoubleTap() {
+    final position = tapDownDetails!.localPosition;
```
Comment (drminh2807) [inline lib/pages/image_viewer/image_viewer.dart:79]: please check null instead
```suggestion
    final position = tapDownDetails?.localPosition;
    if (position == null) return;
```

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart [hunk 11]
```diff
@@ -61,6 +38,42 @@ extension MatrixFileExtension on MatrixFile {
     return;
   }
 
+  void downloadImageInWeb(BuildContext context) {
```
Comment (drminh2807) [inline lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:41]: if it is not mobile, you can use this library instead
https://pub.dev/packages/file_saver

Comment (drminh2807) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:41]: it worked for web, also desktop in the future

Comment (nqhhdev) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart:41]: IMO: No need

File: lib/pages/image_viewer/image_viewer.dart (hunk 8 above)
Comment (imGok) [reply lib/pages/image_viewer/image_viewer.dart:28]: took a snipet, I removed it and still work

Comment (drminh2807) [reply lib/pages/image_viewer/image_viewer.dart:28]: I mean you should create it directly
```suggestion
  final _transformationController = TransformationController();
```

File: lib/utils/matrix_sdk_extensions/matrix_file_extension.dart (hunk 11 above)
Comment (imGok) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart]: great lib, it's better than native download. had to do some extensions to support mimetype tho

Comment (imGok) [reply lib/utils/matrix_sdk_extensions/matrix_file_extension.dart]: 9519e309ee9780034b3d82838f02d9d815f2a81b

**Review by hoangdat — APPROVED** [review]

**Review by drminh2807 — APPROVED** [review]

**Pull request merged**